pip/req/req_set.py

```diff
@@ -288,33 +296,247 @@ def uninstall(self, auto_confirm=False):
+
+        :param constraints: The constraints to choose under. A dict of
+            name -> tuple(tuple(extranames), specifierset)
+        :param chosen: The currently chosen requirements. A dict of
+            name -> tuple(tuple(extranames), version)
+        :param pending: A set of (name, tuple(extranames), specifier)
+            dependencies that have yet to be processed.
+
+        constraints, chosen and pending are not mutated.
+
+        :return: successfully chosen version.
+        :raises: BackTrack if the constraints cannot be solved.
+        """
+        # The current implementation is recursive: the stack contains
+        # the path through the solution space being taken.
+        if not pending:
+            return chosen
+        constraint = frozenset(itertools.islice(pending, 0, 1))
+        new_pending = pending.difference(constraint)
+        name, extras, specifier = list(constraint)[0]
+        logger.debug("Evaluating %s%s %s" % (name, extras, specifier))
+        if name in constraints:
+            co_constraint = constraints[name]
+            specifier &= co_constraint[1]
+            extras = extras + co_constraint[0]
+        if name in chosen:
+            if not specifier.contains(chosen[name][1]):
+                msg = ("version %s of %s not compatible with %s" %
+                       (chosen[name][1], name, specifier))
+                logger.debug("Backtracking: %s" % msg)
+                raise BackTrack(msg, name)
+            extras = extras + chosen[name][0]
+        # the constraint isn't known to be incompatible.
+        # 1) turn it into a constraint for next steps to consider
+        constraints = constraints.copy()
+        constraints[name] = (tuple(sorted(set(extras))), specifier)
+        chosen = chosen.copy()
+        if name in chosen:
+            # We checked above for compat.
+            # Whatever frame added name will do probing in the event
+            # of failures. Merge extras and cast forward.
+            # XXX: Possible memory pressure optimisation: test extras for
+            # difference before copying chosen.
+            chosen[name] = (
+                tuple(sorted(set(extras + chosen[name][0]))), chosen[name][1])
+            return self.resolve_constraints(constraints, chosen, new_pending)
+        # Loop over all possible versions in preference order.
+        last_error = None
+        for candidate in self.req_cache.get_versions(name, self.upgrade):
+            logger.debug("Trying %s %s" % (name, candidate))
+            if not specifier.contains(candidate):
+                # Incompatible with existing constraints.
+                msg = "%s %s incompatible with %s" % (
+                    name, candidate, specifier)
+                logger.debug(msg)
+                continue
+            chosen[name] = (extras, candidate)
+            requires = self.req_cache.requires(name, extras, candidate)
+            try_pending = set(requires)
+            try:
```

Comment:
it's NP-Complete I believe, so they will and they will be bad. On my mental todo list is an iteration counter and a limit (e.g. 1M rounds), plus need to profile a few different styles of immutable data here - e.g. rather than shrinking pending, perhaps build up a completed set, or use a list with careful mutations, or a list of lists or tuple tree.